### PR TITLE
Improve Type Declarations

### PR DESCRIPTION
**Pull request by maxmellen**
This PR aims to improve the type declarations to more accurately reflect the documentation in the README as well as allow compilation in strict mode.

It also includes an updated package-lock.json file as a result of me simply running `npm install`. The only differences in that file seem to be values used by newer versions of NPM.

### CODE REVIEW

**Comment by maxmellen** [comment]
Edit: This PR used to only make lambda-api compatible with TypeScript's strict mode, but after noticing the existing type declarations weren't so accurate, I also included some improvements in this PR and updated the PR's description.

**Comment by maxmellen** [comment]
@jeremydaly Any chance of seeing this merged?

File: index.d.ts [hunk 1]
```diff
@@ -197,10 +204,10 @@ export declare class API {
 
 
 
-  use(path: string, ...middleware: Middleware[]);
-  use(paths: string[], ...middleware: Middleware[]);
-  use (...middleware: Middleware[]);
-  use (...errorHandlingMiddleware: ErrorHandlingMiddleware[]);
+  use(path: string, ...middleware: Middleware[]): void;
+  use(paths: string[], ...middleware: Middleware[]): void;
+  use(...middleware: Middleware[]): void;
```
Comment (jalooc) [inline index.d.ts]: I think you can improve this even further by allowing mixing normal middleware with error-handling middleware, cause if I see correctly the library allows it. Something like this, but prettier of course ;P :
```typescript
type AnyMiddleware = Middleware | ErrorHandlingMiddleware

use(path: string, ...middleware: Middleware[]): void;
use(paths: string[], ...middleware: Middleware[]): void;
use(...middleware: AnyMiddleware[]): void;
```

**Comment by jalooc** [comment]
This PR is essential to use this library with TS, it fixes incorrectly typed basic functionalities. Let's merge it!

File: index.d.ts [hunk 2]
```diff
@@ -36,10 +36,10 @@ export declare interface App {
   [namespace: string]: HandlerFunction;
 }
 
-export declare type Middleware = (req: Request, res: Response, next: Middleware) => void;
+export declare type Middleware = (req: Request, res: Response, next: () => void) => void;
 export declare type ErrorHandlingMiddleware = (error: Error, req: Request, res: Response, next: ErrorHandlingMiddleware) => void;
```
Comment (jalooc) [inline index.d.ts]:
```suggestion
export declare type ErrorHandlingMiddleware = (error: Error, req: Request, res: Response, next: () => void) => void;
```

File: index.d.ts [hunk 3]
```diff
@@ -36,10 +36,10 @@ export declare interface App {
...
 export declare type ErrorCallback = (error?: Error) => void;
-export declare type HandlerFunction = (req?: Request, res?: Response, next?: NextFunction) => void | {} | Promise<{}>;
+export declare type HandlerFunction = (req: Request, res: Response, next?: NextFunction) => void | any | Promise<any>;
 export declare type LoggerFunction = (message: string) => void;
```
Comment (jalooc) [inline index.d.ts]: https://github.com/jeremydaly/lambda-api#adding-additional-detail
```suggestion
 export declare type LoggerFunction = (message: string, additionalDetail?: any) => void;
```

**Comment by jeremydaly** [comment]
@jalooc, let me know when you think this is ready.

**Comment by jalooc** [comment]
I'm not TS-savvy enough to say if it it'd be possible, but it would be very nice to have `req` typed in a smarter way, for a use case like this:

```ts
api.use(async req => {
  req.connection = await createConnection()
  next()
})

// ...

api.finally(req => {
  return req.connection.close() // Can I have the req.connection type automatically inferred?
})
```

Does anybody know if it's possible? If not automatically inferred, then maybe a generic type for `api` could be used? 🤔

**Comment by maxmellen** [comment]
@jalooc You'd want to have the req argument to be typed like `T extends Response` and then you would write your custom type extending `Response` adding the fields you need. You would then cast inside of the handler:
```typescript

interface UserReq extends Response {
  userToken: string
}

api.use(req => {
  const userReq = req as UserReq // TypeScript accepts this cast without any warning, even in the strictest settings
  // do something with userReq.userToken...
})
```

@jalooc @jeremydaly, I'm happy to apply the suggestions, but can we get this merged in a state not so far from how this PR currently stands? It would still be an incremental improvement in the right direction.

Or of course feel free to take over this PR.

**Comment by jalooc** [comment]
@maxmellen , I haven't found any more problems so far, so if you're willing to accept my change suggestions that'd be great. It's a green light from my side. Nice job!

**Comment by jalooc** [comment]
So what we're doing with the discrepancies I listed? Another PR?

**Comment by jeremydaly** [comment]
Yes, please create a new PR.

**Comment by jalooc** [comment]
Just as a reference for others reading this thread and having a similar question as I did:
> I'm not TS-savvy enough to say if it'd be possible, but it would be very nice to have `req` typed in a smarter way, for use cases like this:
> 
> ```ts
> api.use(async req => {
>   req.connection = await createConnection()
>   next()
> })
> 
> // ...
> 
> api.finally(req => {
>   return req.connection.close() // Can I have the req.connection type automatically inferred?
> })
> ```
> 
> Does anybody know if it's possible? If not automatically inferred, then maybe a generic type for `api` could be used? 🤔

@maxmellen The way [you suggested](https://github.com/jeremydaly/lambda-api/pull/111#issuecomment-516833204) is okay in the short run, but after a while of playing with TS, I found a lot of shortcomings. First of all, you'd quickly sink in the depth of overridden interfaces. What's worse, you'd also have to always remember to manually cast `X as Y` to get the right type and that usually means a lot of places in a standard lambda. Besides, using such overridden type would prove problematic when interconneting it with other functionalities from the 3rd party library that don't understand it, so some `as unknown`-type hacks would be necessary.

I found an imho better and **the proper** way to do it: _Declaration merging_ or [_module augmentation_](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) more specifically.

Following the example from my question, this would solve it:
```ts
declare module 'lambda-api' {
  interface Request {
    connection: SomeType,
  }
}
```

and from now on you have a typable `req.connection` everywhere.

**Pull request merged**